File: c/mt76_csi.c
/* SPDX-License-Identifier: BSD-3-Clause-Clear */
/* Protocol facts: gen4m 8fddb9d7, pinned firmware22; docs/STATION_CSI.md.
 * Independent bounded implementation; no firmware code, USB I/O or calibration. */
#include "mt76_csi.h"
#include <string.h>

static uint32_t le32(const uint8_t *p) {
    return (uint32_t)p[0] | (uint32_t)p[1] << 8 | (uint32_t)p[2] << 16 | (uint32_t)p[3] << 24;
}
static unsigned le16(const uint8_t *p) { return p[0] | (unsigned)p[1] << 8; }
static bool address_ok(const uint8_t *p) {
    if (!p || (p[0] & 1)) return false;
    unsigned any = 0;
    for (unsigned i = 0; i < 6; i++) any |= p[i];
    return any != 0;
}
int mt_csi_request(int chip, int action, unsigned receivers, const uint8_t *ta,
                    uint8_t *out, size_t capacity) {
    if (chip != MT_CHIP_MT7925 || action < MT_CSI_STOP || action > MT_CSI_REMOVE_TRANSMITTER || !out)
        return -1;
    if ((receivers && action != MT_CSI_RECEIVER_COUNT) ||
        (ta && action != MT_CSI_ADD_TRANSMITTER && action != MT_CSI_REMOVE_TRANSMITTER)) return -1;
    unsigned len = action < 2 ? 8 : action == 2 ? 15 : action == 3 ? 12 : 16;
    if (capacity < len || (action == 3 && receivers != 1 && receivers != 2) ||
        (action >= 4 && !address_ok(ta))) return -1;
    uint8_t request[16] = {0};
    request[4] = action <= 3 ? (uint8_t)action : 4;
    request[6] = (uint8_t)(len - 4);
    if (action == 2) request[9] = 0x20;
    if (action == 3) request[8] = (uint8_t)receivers;
    if (action >= 4) { request[8] = action == MT_CSI_ADD_TRANSMITTER; memcpy(request + 10, ta, 6); }
    memcpy(out, request, len);
    return (int)len;
}
static int event_body(int chip, const uint8_t *raw, size_t len, unsigned eid, unsigned seq) {
    if (chip != MT_CHIP_MT7925 || !raw || len < 44 || seq > 15) return -1;
    uint32_t word = le32(raw), size = word & 65535;
    if (size < 44 || size > len || word >> 27 != 7 || ((word >> 16) & 15) == 1 ||
        raw[36] != eid || raw[37] != seq) return -1;
    return (int)(size - 44);
}
int mt_csi_ack(int chip, const uint8_t *raw, size_t len, uint8_t seq, uint32_t *status) {
    if (!seq || !status || event_body(chip, raw, len, 1, seq) != 8 || le32(raw + 44) != 0x4a) return -1;
    *status = le32(raw + 48);
    return 0;
}
int mt_beacon_csi_parse(int chip, const uint8_t *raw, size_t len, mt_beacon_csi_report_t *out) {
    int length = event_body(chip, raw, len, 0x4a, 0);
    if (!out || length < 8 || length > 8192) return -1;
    const uint8_t *body = raw + 44;
    if (le16(body + 4) || le16(body + 6) != (unsigned)length - 4) return -1;
    const uint8_t *fields[64] = {0};
    uint32_t sizes[64] = {0};
    unsigned pos = 8, last = 64;
    while (pos < (unsigned)length) {
        if ((unsigned)length - pos == 36 && last == 25 && sizes[25] == 4) {
            unsigned any = 0;
            for (unsigned i = pos; i < (unsigned)length; i++) any |= body[i];
            if (!any) break;
        }
        if ((unsigned)length - pos < 8) return -1;
        unsigned tag = le32(body + pos), size = le32(body + pos + 4);
        pos += 8;
        if (tag > 63 || size > 8192 || fields[tag] || size > (unsigned)length - pos) return -1;
        fields[tag] = body + pos; sizes[tag] = size; last = tag;
        pos += size;
    }
    const unsigned required[] = {0, 1, 2, 3, 4, 5, 8, 9, 12, 18, 20, 21, 25};
    for (unsigned i = 0; i < sizeof(required)/sizeof(*required); i++)
        if (sizes[required[i]] != 4) return -1;
    if (le32(fields[0]) != 22 || le32(fields[5]) != 64 || le32(fields[12]) != (11U << 16 | 1) ||
        le32(fields[18]) > 1 || le32(fields[1]) || le32(fields[4]) || le32(fields[8]) ||
        le32(fields[20]) || le32(fields[21]) || sizes[10] != 8 || sizes[6] != 128 || sizes[7] != 128 ||
        !address_ok(fields[10])) return -1;
    mt_beacon_csi_report_t report = {0};
    report.version = 22; report.data_count = 64; report.rx_index = le32(fields[18]);
    report.rx_mode_raw = 1; report.rx_rate_raw = 11; report.channel_index_raw = le32(fields[9]);
    unsigned rssi = le32(fields[2]) & 255;
    report.rssi_raw_s8 = rssi < 128 ? (int)rssi : (int)rssi - 256;
    report.snr_raw = le32(fields[3]); report.mcu_gpt_raw = le32(fields[25]);
    memcpy(report.transmitter, fields[10], 6);
    for (unsigned i = 0; i < 64; i++) {
        unsigned a = le16(fields[6] + 2*i), b = le16(fields[7] + 2*i);
        report.i[i] = (int16_t)(a < 32768 ? (int)a : (int)a - 65536);
        report.q[i] = (int16_t)(b < 32768 ? (int)b : (int)b - 65536);
    }
    *out = report;
    return 0;
}

File: mt76_csi.py
# SPDX-License-Identifier: BSD-3-Clause-Clear
"""Experimental pinned MT7925 beacon/20MHz CSI wire primitives; no I/O.

See docs/STATION_CSI.md for source, firmware and hardware evidence. These helpers
do not establish a streaming lifecycle, calibrate I/Q, or identify a beacon from
the report alone. Apply the beacon selector and host filtering in an explicitly
owned session. Addresses and coefficient arrays are sensitive in-memory data.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field
from enum import IntEnum


class CsiAction(IntEnum):
    STOP = 0
    START = 1
    BEACON_SELECTOR = 2
    RECEIVER_COUNT = 3
    ADD_TRANSMITTER = 4
    REMOVE_TRANSMITTER = 5


def build_csi_request(chip: str, action: CsiAction, *, receivers=0, transmitter=None) -> bytes:
    """Finite band0 controls only; START clears firmware transmitter filters.

    Apply transmitter allowlist AFTER START, then receiver count LAST. Firmware ACKs
    do not discard already queued events: the host must enforce its filter too.
    STOP does not restore every configuration field; reload after experiments.
    """
    if chip != "mt7925":
        raise ValueError("CSI profile supports MT7925 only")
    if not isinstance(action, CsiAction) or type(receivers) is not int:
        raise ValueError("invalid CSI action or receiver count")
    if receivers != 0 and action != CsiAction.RECEIVER_COUNT:
        raise ValueError("receiver count only applies to its CSI action")
    if transmitter is not None and action not in (
        CsiAction.ADD_TRANSMITTER,
        CsiAction.REMOVE_TRANSMITTER,
    ):
        raise ValueError("transmitter only applies to CSI filter actions")
    if action in (CsiAction.STOP, CsiAction.START):
        return struct.pack("<4xHH", action, 4)
    if action == CsiAction.BEACON_SELECTOR:
        return struct.pack("<4xHHBI2x", 2, 11, 0, 0x20)
    if action == CsiAction.RECEIVER_COUNT:
        if receivers not in (1, 2):
            raise ValueError("CSI requires one or two receivers")
        return struct.pack("<4xHHB3x", 3, 8, receivers)
    if (
        not isinstance(transmitter, bytes)
        or len(transmitter) != 6
        or transmitter[0] & 1
        or not any(transmitter)
    ):
        raise ValueError("CSI filter requires one nonzero unicast transmitter")
    return struct.pack("<4xHHBB6s", 4, 12, action == CsiAction.ADD_TRANSMITTER, 0, transmitter)


def _event_body(chip, raw, eid, sequence):
    if chip != "mt7925" or type(sequence) is not int or not 0 <= sequence <= 15 or len(raw) < 44:
        raise ValueError("unsupported or short CSI event")
    word = struct.unpack_from("<I", raw)[0]
    size = word & 65535
    if (
        not 44 <= size <= len(raw)
        or word >> 27 != 7
        or (word >> 16) & 15 == 1
        or raw[36] != eid
        or raw[37] != sequence
    ):
        raise ValueError("invalid CSI event DMA/type/identity")
    return raw[44:size]


def parse_csi_ack(chip: str, raw: bytes, sequence: int) -> int:
    """Return status only from an exact matched UNI4a command-result envelope."""
    if not sequence:
        raise ValueError("CSI ACK requires nonzero sequence")
    body = _event_body(chip, raw, 1, sequence)
    if len(body) != 8 or struct.unpack_from("<I", body)[0] != 0x4A:
        raise ValueError("invalid CSI command result")
    return struct.unpack_from("<I", body, 4)[0]


def parse_csi_tlvs(body: bytes) -> dict[int, bytes]:
    """Bounded structural parser, also used by broader research-only layouts.

    Unknown tags are length-checked and returned, not interpreted. Do not log
    this dictionary: it contains private transmitter and coefficient data.
    """
    if not 8 <= len(body) <= 8192:
        raise ValueError("CSI body size")
    if struct.unpack_from("<HH", body, 4) != (0, len(body) - 4):
        raise ValueError("CSI outer TLV")
    fields = {}
    pos = 8
    while pos < len(body):
        if (
            len(body) - pos == 36
            and fields
            and next(reversed(fields)) == 25
            and len(fields[25]) == 4
            and not any(body[pos:])
        ):
            break
        if len(fields) >= 64 or len(body) - pos < 8:
            raise ValueError("CSI inner header")
        tag, length = struct.unpack_from("<II", body, pos)
        pos += 8
        if tag > 63 or length > 8192:
            raise ValueError("CSI out-of-range inner header")
        if tag in fields or length > len(body) - pos:
            raise ValueError("CSI duplicate or truncated field")
        fields[tag] = body[pos : pos + length]
        pos += length
    return fields


@dataclass(frozen=True)
class BeaconCsiReport:
    version: int
    data_count: int
    rx_index: int
    tx_index: int
    rx_mode_raw: int
    rx_rate_raw: int
    channel_index_raw: int  # Not an RF channel number; observed zero on channel36.
    rssi_raw_s8: int
    snr_raw: int  # No calibrated SNR conversion.
    mcu_gpt_raw: int  # Wrapping32-bit, approximately1us; NOT TSF/RXD/ToA.
    transmitter: bytes = field(repr=False)
    i: tuple[int, ...] = field(repr=False)
    q: tuple[int, ...] = field(repr=False)


def parse_beacon_csi(chip: str, raw: bytes) -> BeaconCsiReport:
    """Accept only evidenced version22, band0, 20MHz OFDM6, 64-I/Q profile.

    Reject CCK count13/storage64 as unusable, unknown versions, wider/segmented
    reports and unsupported receiver indices. Validity still requires a matching
    configured session/epoch/filter, not just these bytes. No pairing is implied.
    """
    fields = parse_csi_tlvs(_event_body(chip, raw, 0x4A, 0))
    required = (0, 1, 2, 3, 4, 5, 8, 9, 12, 18, 20, 21, 25)
    if any(len(fields.get(tag, b"")) != 4 for tag in required):
        raise ValueError("CSI required scalar size")
    words = {tag: struct.unpack("<I", fields[tag])[0] for tag in required}
    if (
        words[0] != 22
        or words[5] != 64
        or words[12] != (11 << 16 | 1)
        or words[18] not in (0, 1)
        or any(words[tag] for tag in (1, 4, 8, 20, 21))
    ):
        raise ValueError("CSI report outside pinned beacon20 profile")
    if len(fields.get(10, b"")) != 8 or any(len(fields.get(tag, b"")) != 128 for tag in (6, 7)):
        raise ValueError("CSI transmitter or I/Q dimensions")
    transmitter = fields[10][:6]
    if transmitter[0] & 1 or not any(transmitter):
        raise ValueError("CSI invalid transmitter")
    rssi = words[2] & 255
    return BeaconCsiReport(
        22,
        64,
        words[18],
        0,
        1,
        11,
        words[9],
        rssi if rssi < 128 else rssi - 256,
        words[3],
        words[25],
        transmitter,
        struct.unpack("<64h", fields[6]),
        struct.unpack("<64h", fields[7]),
    )
